### ISSUE TITLE
Support ESP-IR

### ISSUE THREAD

**Issue by TheChatty** [gemu2015/Sonoff-Tasmota#40]
SML works fine with some ESP8266 and D0 transceiver on hardware and software serial.
![image](https://user-images.githubusercontent.com/4789510/193695563-797a976d-4612-497c-9093-87aacfe8b94e.png)

Now I'd like to change to ESP-IR (integrated board with emitter and receiver):
![image](https://user-images.githubusercontent.com/4789510/193694627-3a92959c-1742-4bd9-9d47-27bf76d5bf38.png)

- First solution configured to `Serial TX/RX`:  transmitter is always off and only blinking on `SerialSend ...`.
- Second solution configured to `Serial TX/RX`: transmitter is always on and blinking off on `SerialSend ...`. Both diodes are wired to GPIO4/14 thus software serial.

I guess there is an [inverting Schmitt trigger](https://de.wikipedia.org/wiki/Schmitt-Trigger#Invertierender_Schmitt-Trigger) in the [D0 head](https://wiki.volkszaehler.org/hardware/controllers/ir-schreib-lesekopf-ttl-ausgang). For obvious reasons that cannot be easily integrated into ESP-IR.
Is there an easy solution to invert their behavior for SML? (e.g. change all lines in [TasmotaSerial.cpp](https://github.com/arendst/Tasmota/blob/development/lib/default/TasmotaSerial-3.5.0/src/TasmotaSerial.cpp) containg `digitalRead` and `digitalWrite` in a clever way)

**Comment by gemu2015** [comment]
this took me 15 minutes to make a compile option to invert software serial, would be some more effort to make that configurable with a command

#define TS_SWS_INVERT

see repo

**Comment by TheChatty** [comment]
You [forgot one digitalWrite](https://github.com/gemu2015/Sonoff-Tasmota/blob/universal/lib/default/TasmotaSerial-3.5.0/src/TasmotaSerial.cpp#L82).

But even with this fixed the transmitter is always on and flashing when sending. What's happening?

**Comment by gemu2015** [comment]
this is only working for software serial , hardware serial can not be inverted.

the pin i forgot should set the trx GPIO to the right level.

it should definitely work for sending.

for receiving i forgot the level change interrupt active transition, i will check that later

**Comment by TheChatty** [comment]
The problem was `#define`s from `user_config_override.h` are not incorporated into (all?) lib(s).

Now enforcing that `#define  TS_SWS_INVERT` it works visually.

**Comment by gemu2015** [comment]
yes i forgot that we are in a library.
i fixed the irq , try receiving

**Comment by TheChatty** [comment]
You need to check it in ;-)

**Comment by TheChatty** [comment]
Testing receiption with my EasyMeter (it sends telegrams every 2secs) does not work. Need to test later against my IR head in both directions.

**Comment by TheChatty** [comment]
**Sending strings** from ESP-IR with up to 14.400 baud **works**. Starting from 19.200 things get messy. The meter I have which requires sending is only at 1.200 baud thus ok.

**Receiving strings** with ESP-IR **does not work** even at 1.200 or 300 baud. My Kamstrup request (` 80 3f 10 07 00 3c 00 56 00 57 00 59 00 4a 00 44 00 7b 7c 97 0d`) provokes every 10-15 time a single character to be received but mostly none. The character received is in 50% an `@`.
When I undo inversion code for reading (define and interrupt) each Kamstrup request provokes almost steadily receiption of `FF FF`.
Do you have any ideas how to analyze this?

**Comment by gemu2015** [comment]
first check if the non inverse version works both for sending and receiving.
ich checked the code again and can't find a reason why it should not work.

**Comment by TheChatty** [comment]
Using **original** Tasmota on both ends leads to this:

Both modules **configured to IRsend/IRrecv**

- ESP-IR transmitter **off** when idle
- D0 transmitter **lit** when idle

`IRsend {"Protocol":"NEC","Bits":32,"Data":0x1248FEB7}` is received as `"IrReceived":{"Protocol":"UNKNOWN","Bits":32,"Hash":"0x1248FEB7"}` **in both directions**.
This library seems to handle inverted diodes transparently.

Both modules **configured to TCP Tx/Rx with `TCPBaudrate 1200`**

- ESP-IR transmitter **lit** when idle
- D0 transmitter **off** when idle

Data sent from D0 to ESP-IR is **not received**. 
Data sent from ESP-IR to D0 is **kind of received**. Long strings of the same character are "**translated**" into different characters (with some garbage).
```
0x65 -> 0x4D
0x67 -> 0x4C
0x6B -> 0x4A
0x79 -> 0x43
0x31 -> 0x67
0x33 -> 0x66
0x35 -> 0x65
0x3F -> 0x60
0x5F -> 0x50
```

**Comment by TheChatty** [comment]
@s-hadinger Are you able to to give some insight why a bit-inverted TasmotaSerial.cpp installed on ESP-IR is working for sending but not for receiving?Maybe you've got some idea where to continue investigation?

**Comment by s-hadinger** [comment]
This looks like a duplicate from https://github.com/arendst/Tasmota/discussions/16447

**Comment by s-hadinger** [comment]
If that's the same issue, it is possible to invert Hardware Serial, but it requires an additional parameter that is not present in TasmotaSerial (nor in Arduino if I remember well). The standard Tasmota has no plan to introduce this change.

Honestly this does not look as the good approach. IR is not meant to be used as On/Off, as this generates too much interference. It is typically sent with a carrier of 38KHz so that the receiver can filter out glitches and only retain what correpond to the carrier frequency. You can probably use higher frequency carriers, but this requires much more than just plugging serial to the IR diode.

**Comment by TheChatty** [comment]
The discussion at the main repo got stalled so sorry for the duplicate. Once solved I'll close it with a link to the solution.

The IR communication both threads are dealing with is about reading certain energy meters via standardized D0 interface. It's not about signals from TV remotes and so on.

As I said above the D0 IR head probably sports a Schmitt trigger (see linked schematics above) which inverts plain TTL signals thus the head is **working as is** on hardware and software serial from Tasmota or on a FTDI.

The factory-made ESP-IR has its diodes connected to GPIO 4 and 14 thus software serial. As try and error has shown switching HIGH and LOW **did work for sending**. This makes me wonder what is missing for receiving.

**Comment by gemu2015** [comment]
@TheChatty 
sorry i did not realize what you are trying to do.
i just thought you wanted an inverted serial driver to save a hardware inverter.
but the esp-ir you are using is an ir-remote controller which operates with modulated ir light (38kHz). Thus the ir receiver is not just a photodiode but a receiver IC.
it cannot operate with energy meters at all.

**Comment by TheChatty** [comment]
The receiving IC is a [VS838](https://datasheetspdf.com/pdf-file/796250/ETC/VS838/1). Thus it would need to be desoldered and replaced with a [SFH309FA](https://docs.rs-online.com/0d92/0900766b814f56a7.pdf), for instance?

But still strange to me that using IRsend/IRrecv module from Tasmota I can exchange data between ESP-IR (VS838) and D0 head (SFH309FA).

**Comment by gemu2015** [comment]
yes that would work. if you stay with inverted logic you should wire the photodiode so that the logic level is zero when no light is present.

**Comment by TheChatty** [comment]
I'd try to simply connect SFH309FA to the same pin where VS838 is connected at the moment. I assume it's directly hooked up to GPIO14 from ESP. Maybe no bit inversion is need for receiption.
